File: src/main/java/plugin/javafxtools/service/Base64CodecService.java
package plugin.javafxtools.service;

import java.nio.charset.Charset;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/** Text-oriented Base64 codec with explicit alphabet and character encoding. */
public final class Base64CodecService {
    public static final int MAX_INPUT_CHARACTERS = 1_000_000;

    public String encode(String input, Variant variant, TextEncoding encoding) {
        String value = validate(input);
        byte[] bytes = value.getBytes(encoding.charset());
        return switch (variant) {
            case STANDARD -> Base64.getEncoder().encodeToString(bytes);
            case URL_SAFE -> Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
            case MIME -> Base64.getMimeEncoder().encodeToString(bytes);
        };
    }

    public String decode(String input, Variant variant, TextEncoding encoding) {
        String value = validate(input);
        try {
            byte[] decoded = switch (variant) {
                case STANDARD -> Base64.getDecoder().decode(removeWhitespace(value));
                case URL_SAFE -> Base64.getUrlDecoder().decode(removeWhitespace(value));
                case MIME -> decodeMime(value);
            };
            try {
                return encoding.charset().newDecoder()
                        .onMalformedInput(CodingErrorAction.REPORT)
                        .onUnmappableCharacter(CodingErrorAction.REPORT)
                        .decode(java.nio.ByteBuffer.wrap(decoded)).toString();
            } catch (CharacterCodingException e) {
                throw new IllegalArgumentException("解码后的字节不符合所选字符编码", e);
            }
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("输入不是有效的 " + variant.displayName()
                    + " 内容", e);
        }
    }

    private static String validate(String input) {
        String value = input == null ? "" : input;
        if (value.length() > MAX_INPUT_CHARACTERS) {
            throw new IllegalArgumentException("输入不能超过 "
                    + MAX_INPUT_CHARACTERS / 1_000_000 + " 百万字符");
        }
        return value;
    }

    private static String removeWhitespace(String value) {
        StringBuilder normalized = new StringBuilder(value.length());
        value.codePoints().filter(codePoint -> !Character.isWhitespace(codePoint))
                .forEach(normalized::appendCodePoint);
        return normalized.toString();
    }

    private static byte[] decodeMime(String value) {
        String normalized = removeWhitespace(value);
        boolean validAlphabet = normalized.chars().allMatch(character ->
                character >= 'A' && character <= 'Z'
                        || character >= 'a' && character <= 'z'
                        || character >= '0' && character <= '9'
                        || character == '+' || character == '/' || character == '=');
        if (!validAlphabet) {
            throw new IllegalArgumentException("MIME Base64 只能包含编码字符和空白");
        }
        return Base64.getDecoder().decode(normalized);
    }

    public enum Variant {
        STANDARD("标准 Base64"),
        URL_SAFE("URL 安全 Base64"),
        MIME("MIME Base64");

        private final String displayName;

        Variant(String displayName) {
            this.displayName = displayName;
        }

        public String displayName() {
            return displayName;
        }

        @Override
        public String toString() {
            return displayName;
        }
    }

    public enum TextEncoding {
        UTF_8("UTF-8", StandardCharsets.UTF_8),
        GB18030("GB18030 / GBK", Charset.forName("GB18030")),
        UTF_16LE("UTF-16 LE", StandardCharsets.UTF_16LE),
        UTF_16BE("UTF-16 BE", StandardCharsets.UTF_16BE);

        private final String displayName;
        private final Charset charset;

        TextEncoding(String displayName, Charset charset) {
            this.displayName = displayName;
            this.charset = charset;
        }

        public Charset charset() {
            return charset;
        }

        @Override
        public String toString() {
            return displayName;
        }
    }
}

File: src/FxTools.Core/Services/Base64CodecService.cs
using System.Text;
using FxTools.Core.Infrastructure;

namespace FxTools.Core.Services;

public static class Base64CodecService
{
    static Base64CodecService() => Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);

    public static string Encode(string? input, Base64Variant variant, TextEncoding encoding)
    {
        string value = Validate(input);
        byte[] bytes = GetEncoding(encoding).GetBytes(value);
        string encoded = Convert.ToBase64String(
            bytes,
            variant == Base64Variant.Mime
                ? Base64FormattingOptions.InsertLineBreaks
                : Base64FormattingOptions.None);
        return variant == Base64Variant.UrlSafe
            ? encoded.TrimEnd('=').Replace('+', '-').Replace('/', '_')
            : encoded;
    }

    public static string Decode(string? input, Base64Variant variant, TextEncoding encoding)
    {
        string normalized = RemoveWhitespace(Validate(input));
        try
        {
            if (variant == Base64Variant.UrlSafe)
            {
                if (normalized.Any(character =>
                        !(char.IsAsciiLetterOrDigit(character) || character is '-' or '_' or '=')))
                {
                    throw new FormatException();
                }
                normalized = normalized.Replace('-', '+').Replace('_', '/');
                normalized = normalized.TrimEnd('=');
                normalized += (normalized.Length % 4) switch
                {
                    0 => string.Empty,
                    2 => "==",
                    3 => "=",
                    _ => throw new FormatException()
                };
            }
            else if (normalized.Any(character =>
                         !(char.IsAsciiLetterOrDigit(character) || character is '+' or '/' or '=')))
            {
                throw new FormatException();
            }

            byte[] decoded = Convert.FromBase64String(normalized);
            return GetEncoding(encoding).GetString(decoded);
        }
        catch (Exception exception) when (exception is FormatException or DecoderFallbackException)
        {
            throw new ArgumentException(
                $"输入不是有效的 {DisplayName(variant)} 内容，或解码结果不符合所选字符编码。",
                nameof(input),
                exception);
        }
    }

    private static string Validate(string? input)
    {
        string value = input ?? string.Empty;
        if (value.Length > MemoryLimits.Base64InputCharacters)
        {
            throw new ArgumentException("输入不能超过 100 万字符。", nameof(input));
        }
        return value;
    }

    private static string RemoveWhitespace(string value)
    {
        if (!value.Any(char.IsWhiteSpace))
        {
            return value;
        }

        StringBuilder builder = new(value.Length);
        foreach (char character in value)
        {
            if (!char.IsWhiteSpace(character))
            {
                builder.Append(character);
            }
        }
        return builder.ToString();
    }

    private static Encoding GetEncoding(TextEncoding encoding) => encoding switch
    {
        TextEncoding.Utf8 => new UTF8Encoding(false, true),
        TextEncoding.Gb18030 => Encoding.GetEncoding(
            "GB18030", EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback),
        TextEncoding.Utf16LittleEndian => new UnicodeEncoding(false, false, true),
        TextEncoding.Utf16BigEndian => new UnicodeEncoding(true, false, true),
        _ => throw new ArgumentOutOfRangeException(nameof(encoding))
    };

    public static string DisplayName(Base64Variant variant) => variant switch
    {
        Base64Variant.Standard => "标准 Base64",
        Base64Variant.UrlSafe => "URL 安全 Base64",
        Base64Variant.Mime => "MIME Base64",
        _ => throw new ArgumentOutOfRangeException(nameof(variant))
    };
}

public enum Base64Variant
{
    Standard,
    UrlSafe,
    Mime
}

public enum TextEncoding
{
    Utf8,
    Gb18030,
    Utf16LittleEndian,
    Utf16BigEndian
}
